src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,261 @@
+#include "realm/deppart/partitions_gpu_impl.hpp"
+#include <cub/cub.cuh>
+#include "realm/nvtx.h"
+
+namespace Realm {
+
+template <int N, typename T>
+struct RectFieldVolumeOp {
+  __device__ __forceinline__
+  size_t operator()(const Rect<N,T>& r) const {
+    return r.volume();
+  }
+};
+
+// everything in header form:
+template <int N, typename T, typename FT>
+void GPUByFieldMicroOp<N,T,FT>::gpu_populate_bitmasks()
+{
+
+    nvtx_range_push("cuda", "byfield_gpu_populate_bitmasks");
+
+    cudaStream_t stream;
+    CUDA_CHECK(cudaStreamCreate(&stream), stream);
+
+    Memory my_mem = field_data[0].inst.get_location();
+
+    nvtx_range_push("cuda", "build parent and instance entries");
+
+    std::vector<size_t> inst_offsets(field_data.size() + 1);
+    size_t inst_size = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      inst_offsets[i] = inst_size;
+      if (field_data[i].index_space.dense()) {
+        inst_size += 1;
+      } else {
+        // only call get_entries() once per source
+        inst_size += field_data[i].index_space.sparsity.impl()->get_entries().size();
+      }
+    }
+    // final end offset
+    inst_offsets[field_data.size()] = inst_size;
+
+    RegionInstance inst_entries_instance = this->realm_malloc(inst_size * sizeof(SparsityMapEntry<N,T>), my_mem);
+    SparsityMapEntry<N,T>* d_inst_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(inst_entries_instance, 0).base);
+
+    size_t inst_pos = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      const IndexSpace<N,T> &inst_space = field_data[i].index_space;
+      if (inst_space.dense()) {
+        // just one rect
+        SparsityMapEntry<N,T> entry;
+        entry.bounds = inst_space.bounds;
+        CUDA_CHECK(cudaMemcpyAsync(d_inst_entries + inst_pos, &entry, sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+        ++inst_pos;
+      } else {
+        auto tmp = inst_space.sparsity.impl()->get_entries();
+        CUDA_CHECK(cudaMemcpyAsync(d_inst_entries + inst_pos, tmp.data(), tmp.size() * sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+        inst_pos += tmp.size();
+      }
+    }
+
+    span<SparsityMapEntry<N,T>> parent_entries;
+    std::vector<SparsityMapEntry<N,T>> parent_entries_vec;
+    RegionInstance parent_entries_instance;
+    SparsityMapEntry<N,T>* d_parent_entries;
+    if (parent_space.dense()) {
+      SparsityMapEntry<N,T> entry;
+      entry.bounds = parent_space.bounds;
+      parent_entries_vec = {entry};
+      parent_entries = span<SparsityMapEntry<N,T>>(parent_entries_vec.data(), 1);
+    } else {
+      parent_entries = parent_space.sparsity.impl()->get_entries();
+    }
+    parent_entries_instance = this->realm_malloc(parent_entries.size() * sizeof(SparsityMapEntry<N,T>), my_mem);
+    d_parent_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(parent_entries_instance, 0).base);
```

Comment:
You could use these https://github.com/StanfordLegion/realm/blob/main/src/realm/instance.h#L133 methods probably if the AffineAccessor is cumbersome.